gdnative-sys/build.rs

```diff
@@ -76,304 +77,304 @@ mod header_binding {
+            impl GodotApi {
+                #impl_constructor
+                #impl_api_functions
+            }
+        };
+        let mut wrapper_file = File::create(to.join(file_name)).expect(&format!(
+            "Couldn't create output file: {:?}",
+            to.join(file_name)
+        ));
+        write!(wrapper_file, "{}", wrapper).unwrap();
+    }
 
-        let is_const = text.starts_with("const ");
+    fn godot_api_struct_fields(api: &ApiRoot) -> TokenStream {
+        api.all_apis()
+            .iter()
+            .map(|api| {
+                let field = api.macro_ident();
+                let godot_api_struct = api.godot_api_struct();
+                quote!(#field: #godot_api_struct,)
+            })
+            .collect()
+    }
 
-        if is_const {
-            text = ty.trim_start_matches("const ");
+    fn api_constructor(api: &ApiRoot) -> TokenStream {
+        let mut godot_api_struct_fields = TokenStream::new();
+        for api in api.all_apis() {
+            let field = api.macro_ident();
+            let gd_api_type = api.godot_api_type();
+            let v_maj = api.version.major;
+            let v_min = api.version.minor;
+            let gd_api_struct = api.godot_api_struct();
+            godot_api_struct_fields.extend(quote!{
+                #field: *(find_api_ptr(core_api_struct, #gd_api_type, #v_maj, #v_min) as *const #gd_api_struct),
+            });
         }
+        quote! {
+            pub unsafe fn from_raw(core_api_struct: *const godot_gdnative_core_api_struct) -> Self {
+                GodotApi{
+                    #godot_api_struct_fields
+                }
+            }
+        }
+    }
 
-        let is_double_pointer = text.ends_with("**");
-        let is_single_pointer = text.ends_with("*");
-
-        if is_double_pointer {
-            text = text.trim_end_matches(" **");
-            text = text.trim_end_matches("**");
-        } else if is_single_pointer {
-            text = text.trim_end_matches(" *");
-            text = text.trim_end_matches("*");
-        } else {
+    fn api_wrapper_functions(api: &ApiRoot) -> TokenStream {
+        let mut result = TokenStream::new();
+        for api in api.all_apis() {
+            for function in &api.functions {
+                let function_name = function.rust_name();
+                let args = function.rust_args_with_types();
+                let return_type = function.rust_return_type();
+                let api_struct_field = api.macro_ident();
+                let arg_names = function.rust_args_without_types();
+                let expect_msg = format!(
+                    "API function missing: {}.{}",
+                    api.godot_api_struct(),
+                    function_name
+                );
+                result.extend(quote! {
+                    pub unsafe fn #function_name(&self, #args) -> #return_type {
+                        self.#api_struct_field.#function_name.expect(#expect_msg)(#arg_names)
```

Comment:
This means that `expect` will be called for each invocation, instead of once at API struct initialization.

Reply:
I agree it would be better to call `expect` many times initially rather that once for each function call. This way if one of the `expect`s is going to fail, it will fail quickly and at a predictable time.